File: 121-BestTimeToBuyAndSellStock/121-BestTimeToBuyAndSellStock.cpp
// Last updated: 16/10/2025, 04:29:20
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int maxProfit = 0;
        int minPrice = prices[0];
        for(int i=0; i<n; i++){
            int profit = prices[i] - minPrice;
            if(profit > maxProfit){
                maxProfit = profit;
            }
            minPrice = min(minPrice,prices[i]);
        }
        return maxProfit;
    }
};

File: 121-BestTimeToBuyAndSellStock/121-BestTimeToBuyAndSellStock.java
// Last updated: 16/10/2025, 04:29:15
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;

        for(int i = 0; i < n; i++){
            if(prices[i] < minPrice){
                minPrice = prices[i]; //Buy at lower price
            }else{
                maxProfit = Math.max(maxProfit, prices[i] - minPrice); //sell at higher price
            }
        }
        return maxProfit;
    }
}
